torchvision/utils.py

```diff
@@ -1,10 +1,13 @@
+from typing import Union, Optional, Sequence, Tuple
+import io
 import torch
 import math
 irange = range
 
 
-def make_grid(tensor, nrow=8, padding=2,
-              normalize=False, range=None, scale_each=False, pad_value=0):
+def make_grid(tensor: Union[torch.Tensor, Sequence[torch.Tensor]], nrow: int = 8,
+              padding: int = 2, normalize=False, range: Optional[Tuple[int, int]] = None,
+              scale_each: bool = False, pad_value: int = 0) -> torch.Tensor:
```

Comment:
Can we somehow make the output type dependent on the input type, i.e. if I pass a `torch.DoubleTensor` I should also get back a `torch.DoubleTensor`?

Reply:
I would have expected it to be the case (except for integer tensors), maybe it's an oversight in the code?
I saw that we use the type of the `tensor` in `grid = tensor.new_full(...)`, so it should keep the original `dtype` normally

Reply:
I think you misunderstood me. If a function returns the same type as its input, it is my understanding that this is the most generic way to annotate it:

```python
from typing import TypeVar

T = TypeVar("T")


def foo(bar: T) -> T:
    bar.baz()
    return bar
```

Here `bar` could be `torch.FloatTensor`, `torch.ByteTensor` and so on. I don't know if it is possible to limit `T` to `torch.Tensor` and all of its subclasses. Depending on the complexity of the annotations I'm not sure if we even want this.

Reply:
I think annotating it as `torch.Tensor` is enough -- we only have a single type of Tensor now, and `FloatTensor` / `DoubleTensor` are legacy and alias to the same `Tensor` object.